Prettier formatting for ConvertTo-Json output. #2736

kittholland opened the pull request:
This change standardizes JSON output to example given, as well as
codemaid and online lint tools.

Sample object used for testing:

```powershell
@{
foo = @{
first = 'a'
second = 'bbbbbbbb'
}
barbarbarbar = @{
first = 'a'
second = 'bbbbbbbb'
NestedArray = @(
'Test3'
'Test4'
'Test5'
3
4
)
NestedObject = @{
MoreObject = 'AnotherObject'
TestBool = $true
}
}
array = @(
'Thing1'
'Thing2'
)
dan = 15
} | ConvertTo-Json
```

kittholland committed:
Prettier formatting for ConvertTo-Json output. #2736

This change standardizes JSON output to example given, as well as
codemaid and online lint tools.

Sample object used for testing:

@{
foo = @{
first = 'a'
second = 'bbbbbbbb'
}
barbarbarbar = @{
first = 'a'
second = 'bbbbbbbb'
NestedArray = @(
'Test3'
'Test4'
'Test5'
3
4
)
NestedObject = @{
MoreObject = 'AnotherObject'
TestBool = $true
}
}
array = @(
'Thing1'
'Thing2'
)
dan = 15
} | ConvertTo-Json

msftclas commented:
Hi __@kittholland__, I'm your friendly neighborhood Microsoft Pull Request Bot (You can call me MSBOT). Thanks for your contribution!
    <span>You've already signed the contribution license agreement. Thanks!</span>
        <p>The agreement was validated by Microsoft and real humans are currently evaluating your PR.</p>

TTYL, MSBOT;

SteveL-MSFT reviewed (changes requested):
On CoreCLR, I would prefer to use the NewtonSoft Json serializer with Format=Formatting.Indented.  For FullCLR, I would just leave it as-is today.  The compiler should leave out all the custom formatting code for a CoreCLR build, otherwise we can `#ifdef` it.

SteveL-MSFT commented:
We should have a test for pretty print.

kittholland committed:
Updated CoreCLR implementation to use NewtonSoft Indented Formatting

I did not change the FullCLR behavior, I was not sure if you meant to
revert my changes or to leave it as is in the current pull request.

SteveL-MSFT reviewed (approved):
LGTM

kittholland commented:
I will work on the pretty print test soon.

kittholland committed:
Added tests that validate pretty Json output.

Not sure if there is a better thought on how to implement these. The
first two fail against current master, but succeed once this PR is
applied. Third test is successful prior and post this PR.

SteveL-MSFT reviewed (approved):
LGTM

SteveL-MSFT commented:
@daxian-dbw the CI breaks are related to https://github.com/PowerShell/PowerShell/pull/2806, how to proceed?

JamesWTruher commented on test/powershell/Modules/Microsoft.PowerShell.Utility/ConvertTo-Json.Tests.ps1:
```diff
@@ -0,0 +1,14 @@
+Describe "ConvertTo-Json" -tags "CI" {
```
>Describe [](start = 0, length = 8)

please add these to test\powershell\Modules\Microsoft.PowerShell.Utility\pester.commands.cmdlets.json.tests.ps1 rather than creating a new file

kittholland replied:
Checking in fix now, thanks for pointing out the correct location.

kittholland committed:
Moved tests and removed extraneous file.

Moved pretty/compressed json tests from standalone file into the
existing ConvertTo-Json test file.

daxian-dbw commented:
@SteveL-MSFT I just restarted the Travis CI build.

kittholland commented:
Thanks for rerunning Travis-CI, I see my tests are failing in that environment. I will have to build out a non windows environment for more testing.

There are also some here-string based tests that are labeled -pending$isCoreCLR and are tagged in a Feature describe block. I think I will need to rework any here-string based tests to accomodate the new formatting.

SteveL-MSFT commented:
Closing and reopening to restart CI

msftclas commented:
Hi __@kittholland__, I'm your friendly neighborhood Microsoft Pull Request Bot (You can call me MSBOT). Thanks for your contribution!
    <span>You've already signed the contribution license agreement. Thanks!</span>
        <p>The agreement was validated by Microsoft and real humans are currently evaluating your PR.</p>

TTYL, MSBOT;

kittholland committed:
Updated tests for cross-platform support

kittholland commented:
Can we retry the travis CI? I have successful test results on ubuntu 14.04.

SteveL-MSFT commented:
Closing and reopening to restart CI

msftclas commented:
Hi __@kittholland__, I'm your friendly neighborhood Microsoft Pull Request Bot (You can call me MSBOT). Thanks for your contribution!
    <span>You've already signed the contribution license agreement. Thanks!</span>
        <p>The agreement was validated by Microsoft and real humans are currently evaluating your PR.</p>

TTYL, MSBOT;

kittholland commented:
Due to differences in newline characters between platforms, I think doing comparisons on here-strings will have to be very carefully handled if it is possible.

I receive a different string hashcode on the output of ConvertTo-Json on linux vs. windows for example.

mirichmo commented:
@SteveL-MSFT and @JamesWTruher This looks like a forgotten PR that is a candidate for merging. Can you please take a second look at it?

SteveL-MSFT reviewed (approved):
LGTM
